src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/helper/builders/NonAssociation.kt

```diff
@@ -0,0 +1,122 @@
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.repository.OffenderNonAssociationRepository
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.repository.ReferenceCodeRepository
+import java.time.LocalDate
+
+@DslMarker
+annotation class NonAssociationDslMarker
+
+@NomisDataDslMarker
+interface NonAssociationDsl {
+  @NonAssociationDetailDslMarker
+  fun nonAssociationDetail(
+    typeSeq: Int = 1,
+    nonAssociationReason: String,
+    recipNonAssociationReason: String? = null,
+    nonAssociationType: String,
+    effectiveDate: LocalDate,
+    expiryDate: LocalDate? = null,
+    authorisedBy: String? = null,
+    comment: String? = null,
+  ): OffenderNonAssociationDetail
+}
+
+@Component
+class NonAssociationBuilderRepository(
+  private val nonAssociationReasonRepository: ReferenceCodeRepository<NonAssociationReason>? = null,
+  private val nonAssociationTypeRepository: ReferenceCodeRepository<NonAssociationType>? = null,
+  private val nonAssociationRepository: OffenderNonAssociationRepository? = null,
+) {
+  fun lookupNonAssociationReason(code: String): NonAssociationReason =
+    nonAssociationReasonRepository?.findByIdOrNull(ReferenceCode.Pk(NonAssociationReason.DOMAIN, code))!!
+
+  fun lookupNonAssociationType(code: String): NonAssociationType =
+    nonAssociationTypeRepository?.findByIdOrNull(ReferenceCode.Pk(NonAssociationType.DOMAIN, code))!!
+
+  fun save(nonAssociation: OffenderNonAssociation) =
+    nonAssociationRepository?.findByIdOrNull(nonAssociation.id) ?: nonAssociationRepository?.save(nonAssociation)
+}
+
+@Component
+class NonAssociationBuilderFactory(
+  private val repository: NonAssociationBuilderRepository? = null,
+  private val nonAssociationDetailBuilderFactory: NonAssociationDetailBuilderFactory = NonAssociationDetailBuilderFactory(),
+) {
+  fun builder(): NonAssociationBuilder {
+    return NonAssociationBuilder(
+      repository,
+      nonAssociationDetailBuilderFactory,
+    )
+  }
+}
+
+class NonAssociationBuilder(
+  val repository: NonAssociationBuilderRepository? = null,
+  val nonAssociationDetailBuilderFactory: NonAssociationDetailBuilderFactory,
+) : NonAssociationDsl {
+
+  private lateinit var nonAssociation: OffenderNonAssociation
+
+  fun build(
+    offender: Offender,
+    nsOffender: Offender,
+    offenderBooking: OffenderBooking,
+    nsOffenderBooking: OffenderBooking,
+    nonAssociationReason: String,
+    recipNonAssociationReason: String,
+  ): OffenderNonAssociation =
+    OffenderNonAssociation(
+      OffenderNonAssociationId(
+        offender = offender,
+        nsOffender = nsOffender,
+      ),
+      offenderBooking = offenderBooking,
+      nsOffenderBooking = nsOffenderBooking,
+      nonAssociationReason = repository?.lookupNonAssociationReason(nonAssociationReason),
+      recipNonAssociationReason = repository?.lookupNonAssociationReason(recipNonAssociationReason),
+    )
+      .let { save(it) }
```

Comment:
I don't think you need to save here and the transaction takes care of it.

Reply:
I think this save is persisting a new entity.